### PR TITLE
Introduce job payloads

### PR DESCRIPTION
**Pull request by jakubkuc**
I want to merge this change because...

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-3583 Introduce EventPayloads for Webhooks tasks](https://app.clickup.com/t/2549495/SALEOR-3583)

File: saleor/core/models.py [hunk 1]
```diff
@@ -120,3 +120,23 @@ class Job(models.Model):
...
+    created_at = models.DateTimeField(auto_now_add=True)
+    event_payload = models.ForeignKey(
+        EventPayload,
+        related_name="event_tasks",
+        null=True,
+        on_delete=models.SET_NULL,
+    )
+    task_id = models.CharField(max_length=254)
+    status = models.CharField(max_length=50, choices=JobStatus.CHOICES)
+    error = models.CharField(null=True, blank=True, max_length=254)
```
Comment (maarcingebala) [inline saleor/core/models.py:139]: Should be `response = TextField()`.

File: saleor/plugins/webhook/plugin.py [hunk 2]
```diff
@@ -150,118 +184,164 @@ def draft_order_deleted(self, order: "Order", previous_value: Any) -> Any:
...
         if not self.active:
             return previous_value
-        page_data = generate_page_payload(page)
-        trigger_webhooks_for_event.delay(WebhookEventType.PAGE_DELETED, page_data)
+        page_data = json.dumps(generate_page_payload(page))
+        event_payload = EventPayload.objects.create(payload=page_data)
+
+        trigger_webhooks_for_event.delay(
+            WebhookEventType.PAGE_DELETED, event_payload.id
+        )
```
Comment (maarcingebala) [inline saleor/plugins/webhook/plugin.py:344]: `trigger_webhooks_for_event` doesn't have to be a Celery task.

File: saleor/plugins/webhook/tasks.py [hunk 3]
```diff
@@ -158,96 +175,158 @@ def send_webhook_using_google_cloud_pubsub(
...
+            return WebhookResponse(content=str(e), status=EventDeliveryStatus.FAILED)
+    raise ValueError("Unknown webhook scheme: %r" % (parts.scheme,))
+
+
+@app.task(
+    bind=True,
+    retry_backoff=10,
+    retry_kwargs={"max_retries": 5},
+)
+def send_webhook_request(self, event_delivery_id):
```
[the remark this answers is not in the document]
Comment (maarcingebala) [reply saleor/plugins/webhook/tasks.py:238]: We will add old tasks to this PR, once SQS in Cloud is drained we will remove old tasks.

File: saleor/plugins/webhook/plugin.py [hunk 4]
```diff
@@ -417,3 +394,14 @@ def process_payment(
             previous_value,
             **kwargs,
         )
+
+    # @staticmethod
```
Comment (korycins) [inline saleor/plugins/webhook/plugin.py:398]: 🚔 🚔 🚔

File: saleor/plugins/webhook/tasks.py [hunk 5]
```diff
@@ -60,37 +79,38 @@ def _get_webhooks_for_event(event_type, webhooks=None):
...
+    for delivery in deliveries:
+        send_webhook_request.delay(delivery.id)
 
 
 def trigger_webhook_sync(event_type: str, data: str, app: "App"):
     """Send a synchronous webhook request."""
-    webhooks = _get_webhooks_for_event(event_type, app.webhooks.all())
-    webhook = webhooks.first()
-    if not webhook:
+    webhooks = _get_webhooks_for_event(event_type, app.webhooks.all())[:1]
```
Comment (korycins) [inline saleor/plugins/webhook/tasks.py:96]: Question: Wouldn't be more readable to create EventDelivery here directly? I mean, from one side it is great that we have a single function for that with a bulk but from the other, we have to modify the input and output arguments to operate over a list - like list with single elements, plural name of the variables

File: saleor/plugins/webhook/tasks.py [hunk 6]
```diff
@@ -158,96 +185,172 @@ def send_webhook_using_google_cloud_pubsub(
...
+                "[Webhook] Failed request to %r: %r.",
+                webhook.target_url,
+                response.content,
+            )
             try:
                 countdown = self.retry_backoff * (2 ** self.request.retries)
                 self.retry(countdown=countdown, **self.retry_kwargs)
             except MaxRetriesExceededError:
                 task_logger.warning(
                     "[Webhook] Failed request to %r: exceeded retry limit.",
```
Comment (korycins) [inline saleor/plugins/webhook/tasks.py:274]: Can we add here more details related to wehbook?  Something similiar as is added in line 280-283
ID, event types

File: saleor/plugins/webhook/tasks.py [hunk 7]
```diff
@@ -158,96 +185,172 @@ def send_webhook_using_google_cloud_pubsub(
...
+                webhook.target_url,
+                domain,
+                webhook.secret_key,
+                delivery.event_type,
+                data,
+            )
+        attempt_update(attempt, response)
+        if response.status == EventDeliveryStatus.FAILED:
+            task_logger.info(
+                "[Webhook] Failed request to %r: %r.",
```
Comment (korycins) [inline saleor/plugins/webhook/tasks.py:265]: Can we add here more details related to wehbook?  Something similiar as is added in line 280-283
ID, event type

File: saleor/plugins/webhook/tasks.py [hunk 8]
```diff
@@ -158,96 +185,172 @@ def send_webhook_using_google_cloud_pubsub(
...
                     signature,
-                    event_type,
+                    delivery.event_type,
                     timeout=WEBHOOK_SYNC_TIMEOUT,
                 )
-                response_data = response.json()
+                response_data = json.loads(response.content)
         except RequestException as e:
-            logger.warning("[Webhook] Failed request to %r: %r.", target_url, e)
+            logger.warning(
```
Comment (korycins) [inline saleor/plugins/webhook/tasks.py:323]: IMHO, adding webhook ID would be usefull here

File: saleor/core/tasks.py [hunk 9]
```diff
@@ -1,8 +1,20 @@
...
 from ..celeryconf import app
+from ..settings import EVENT_PAYLOAD_DELETE_PERIOD
+from .models import EventDelivery, EventDeliveryAttempt, EventPayload
 
 
 @app.task
 def delete_from_storage_task(path):
     default_storage.delete(path)
+
+
```
Comment (fowczarek) [inline saleor/core/tasks.py:13]: Should we add `@app.task` here?

File: saleor/settings.py [hunk 10]
```diff
@@ -486,8 +486,16 @@ def get_host():
...
         "schedule": timedelta(days=1),
     },
+    "delete-outdated-event-data": {
+        "task": "saleor.core.tasks.delete_event_payloads_task",
+        "schedule": timedelta(days=1),
+    },
 }
 
+EVENT_PAYLOAD_DELETE_PERIOD = timedelta(
+    seconds=parse(os.environ.get("EVENT_PAYLOAD_DELETE_PERIOD", "7 days"))
```
Comment (fowczarek) [inline saleor/settings.py:496]: IMO we should store payloads at least 2 weeks.
```suggestion
    seconds=parse(os.environ.get("EVENT_PAYLOAD_DELETE_PERIOD", "14 days"))
```

**Review by fowczarek — APPROVED** [review]

File: saleor/core/tasks.py (hunk 9 above)
Comment (maarcingebala) [reply saleor/core/tasks.py:13]: Yes this was a task before and it should be a task.

**Review by maarcingebala — APPROVED** [review]

File: saleor/plugins/webhook/plugin.py [hunk 11]
```diff
@@ -217,66 +247,71 @@ def product_variant_deleted(
...
+            event_payload, WebhookEventType.TRANSLATION_CREATED
         )
 
     def translation_updated(self, translation: "Translation", previous_value: Any):
         if not self.active:
             return previous_value
         translation_data = generate_translation_payload(translation)
-        trigger_webhooks_for_event.delay(
-            WebhookEventType.TRANSLATION_UPDATED, translation_data
+        event_payload = EventPayload.objects.create(payload=translation_data)
```
[the remark this answers is not in the document]
Comment (jakubkuc) [reply saleor/plugins/webhook/plugin.py]: done

File: saleor/plugins/webhook/tasks.py (hunk 5 above)
Comment (jakubkuc) [reply saleor/plugins/webhook/tasks.py]: Yes I agree. fixed it

File: saleor/plugins/webhook/tasks.py (hunk 6 above)
Comment (jakubkuc) [reply saleor/plugins/webhook/tasks.py]: added

File: saleor/plugins/webhook/tasks.py (hunk 7 above)
Comment (jakubkuc) [reply saleor/plugins/webhook/tasks.py]: added

File: saleor/plugins/webhook/tasks.py (hunk 8 above)
Comment (jakubkuc) [reply saleor/plugins/webhook/tasks.py:332]: added more info

**Pull request merged**